src/ape/api/query.py

```diff
@@ -0,0 +1,100 @@
+from typing import Any, Dict, List, Literal, Optional, Union
+
+import pandas as pd
+from pydantic import BaseModel, PositiveInt, root_validator
+
+from ape.managers.networks import NetworkManager
+from ape.types import AddressType
+from ape.utils import abstractdataclass, abstractmethod
+
+QueryType = Union["BlockQuery", "AccountQuery", "ContractEventQuery", "ContractMethodQuery"]
+
+
+class _BaseQuery(BaseModel):
+    type: str  # Used as discriminator
+    columns: List[str]
+
+
+class _BaseBlockQuery(_BaseQuery):
+    start_block: PositiveInt = 0
+    stop_block: Optional[PositiveInt] = None
+
+    @root_validator(pre=True)
+    def check_start_block_before_stop_block(cls, values):
+        if values["stop_block"] < values["start_block"]:
+            raise ValueError(
+                f"stop_block: {values['stop_block']} cannot be less than "
+                f"start_block: {values['start_block']}."
+            )
+
+        return values
+
+
+class BlockQuery(_BaseBlockQuery):
+    type: Literal["blocks"]
+
+
+class _BaseAccountQuery(BaseModel):
+    start_nonce: PositiveInt = 0
+    stop_nonce: Optional[PositiveInt] = None
+
+    @root_validator(pre=True)
+    def check_start_nonce_before_stop_nonce(cls, values):
+        if values["stop_nonce"] < values["start_nonce"]:
+            raise ValueError(
+                f"stop_nonce: {values['stop_nonce']} cannot be less than "
+                f"start_nonce: {values['start_nonce']}."
+            )
+
+        return values
+
+
+class AccountQuery(_BaseAccountQuery):
+    type: Literal["accounts"]
+    account: AddressType
+
+
+class ContractEventQuery(_BaseBlockQuery):
+    type: Literal["contract_events"]
+    contract: AddressType
+    event_id: bytes
+
+
+class ContractMethodQuery(_BaseBlockQuery):
+    type: Literal["contract_calls"]
+    contract: AddressType
+    method_id: bytes
+    method_args: Dict[str, Any]
+
+
+@abstractdataclass
+class QueryAPI:
+
+    network_manager: NetworkManager
+
+    @abstractmethod
+    def estimate_query(self, query: QueryType) -> Optional[int]:
+        """
+        Estimation of time needed to complete the query. The estimation is returned
+        as an int representing milliseconds. A value of None indicates that the
```

Comment:
Is the `int` still milliseconds or is it more like a priority number? Having a hard time picturing how we can know how many milliseconds, even approximately, a query will take

Reply:
The estimate is loose. It definitely takes less than 1 second to query a single block via a Provider, so take a loose estimate of what that is (maybe it's 50ms) and multiply that by number of blocks (let's say you want to get 100 blocks, so that's 5 seconds total)

It doesn't have to be super accurate, just a good estimate to use for comparison's sake